mobx_codegen/lib/src/mobx_codegen_base.dart

```diff
@@ -24,34 +24,43 @@ class StoreGenerator extends Generator {
 
   @override
   FutureOr<String> generate(LibraryReader library, BuildStep buildStep) {
-    Iterable<String> generate(ClassElement baseClass) sync* {
-      final mixedClass =
-          library.classes.where((c) => c != baseClass).firstWhere((c) {
-        if (c.supertype.typeArguments.isNotEmpty &&
-            baseClass.typeParameters.length ==
-                c.supertype.typeArguments.length) {
-          final t = baseClass.type.instantiate(c.supertype.typeArguments);
-          return t.isSupertypeOf(c.type);
-        }
-        return c.supertype == baseClass.type;
-      }, orElse: () => null);
-
-      if (mixedClass != null) {
-        yield generateStoreClassCode(library, baseClass, mixedClass);
-      }
-    }
-
     return library.classes
         .where(isValidStoreClass)
-        .expand(generate)
+        .expand((c) => expandToMixinCode(library, c))
         .toSet()
         .join('\n\n');
   }
 
+  Iterable<String> expandToMixinCode(
+    LibraryReader library,
+    ClassElement baseClass,
+  ) sync* {
+    // Find all classes other than the base class candidate we're evaluating
+    final otherClasses = library.classes.where((c) => c != baseClass);
+    // Attempt to find the first class that extend our base class
+    final mixedClass = otherClasses.firstWhere((c) {
+      // If our base class has different type parameterization requirements than
+      // the class we're evaluating provides, we know it's not a subclass.
+      if (baseClass.typeParameters.length != c.supertype.typeArguments.length) {
+        return false;
+      }
+
+      // Apply the subclass' type arguments to the base type, and perform a
+      // supertype check.
+      return baseClass.type
+          .instantiate(c.supertype.typeArguments)
```

Comment:
Just had a thought — this might(?) throw if the type parameters impose constraints.